File: extension/lib/prompt-parser.js
const BLOCK_SEPARATOR = "|||";
const HEADER_RE = /^\s*0*([1-9][0-9]*)\s*:\s*(.*)\s*$/;
const REF_RE = /@([A-Za-z0-9_-]+)/g;

function normalizeBody(body) {
  return String(body || "")
    .split(/\r?\n/)
    .map((line) => line.replace(/\s+$/, ""))
    .join("\n")
    .trim();
}

function buildTag(number, prefix = "S", padWidth = 3) {
  return `${prefix}${String(number).padStart(Math.max(3, padWidth), "0")}`;
}

export function parsePromptBlocks(rawText, options = {}) {
  const prefix = options.prefix || "S";
  const padWidth = Number(options.padWidth || 3);
  const chunks = String(rawText || "")
    .split(BLOCK_SEPARATOR)
    .map((part) => part.trim())
    .filter(Boolean);

  const items = [];
  for (const chunk of chunks) {
    const lines = chunk.split(/\r?\n/);
    const firstLine = String(lines[0] || "").trim();
    let number = null;
    let body = "";

    const inline = HEADER_RE.exec(firstLine);
    if (inline) {
      number = Number(inline[1]);
      const inlineBody = String(inline[2] || "").trim();
      const restBody = lines.slice(1).join("\n").trim();
      body = normalizeBody([inlineBody, restBody].filter(Boolean).join("\n"));
    } else {
      const onlyHeader = /^\s*0*([1-9][0-9]*)\s*:\s*$/.exec(firstLine);
      if (!onlyHeader) {
        continue;
      }
      number = Number(onlyHeader[1]);
      body = normalizeBody(lines.slice(1).join("\n"));
    }
    if (!number || !body) {
      continue;
    }

    const tag = buildTag(number, prefix, padWidth);
    const referenceNames = [];
    let match;
    while ((match = REF_RE.exec(body)) !== null) {
      referenceNames.push(match[1]);
    }
    REF_RE.lastIndex = 0;

    items.push({
      number,
      tag,
      body,
      raw: chunk,
      referenceNames,
      renderedPrompt: `${tag} Prompt : ${body}`,
      cleanedBody: body.replace(REF_RE, "$1")
    });
  }

  items.sort((a, b) => a.number - b.number);
  return items;
}

export function parseManualNumbers(raw) {
  const wanted = new Set();
  for (const part of String(raw || "").replace(/\s+/g, "").split(",")) {
    if (!part) {
      continue;
    }
    if (part.includes("-")) {
      const [left, right] = part.split("-", 2);
      const lo = Number(left);
      const hi = Number(right);
      if (Number.isFinite(lo) && Number.isFinite(hi) && lo > 0 && hi > 0) {
        const start = Math.min(lo, hi);
        const end = Math.max(lo, hi);
        for (let value = start; value <= end; value += 1) {
          wanted.add(value);
        }
      }
      continue;
    }
    const value = Number(part);
    if (Number.isFinite(value) && value > 0) {
      wanted.add(value);
    }
  }
  return [...wanted].sort((a, b) => a - b);
}

export function filterPromptBlocks(items, settings) {
  const mode = settings.numberMode || "range";
  if (mode === "manual") {
    const wanted = new Set(parseManualNumbers(settings.manualNumbers || ""));
    return items.filter((item) => wanted.has(item.number));
  }
  const start = Number(settings.startNumber || 1);
  const end = Number(settings.endNumber || start || 1);
  const lo = Math.min(start, end);
  const hi = Math.max(start, end);
  return items.filter((item) => item.number >= lo && item.number <= hi);
}

export function summarizeSelection(items) {
  if (!items.length) {
    return "선택된 작업 없음";
  }
  const preview = items.slice(0, 8).map((item) => item.tag).join(", ");
  if (items.length > 8) {
    return `${items.length}개 선택: ${preview} 외 ${items.length - 8}개`;
  }
  return `${items.length}개 선택: ${preview}`;
}

export function compressNumbers(numbers, prefix = "S") {
  const unique = [...new Set(numbers.map((value) => Number(value)).filter((value) => Number.isFinite(value) && value > 0))].sort((a, b) => a - b);
  if (!unique.length) {
    return "";
  }
  const parts = [];
  let start = unique[0];
  let prev = unique[0];
  for (const value of unique.slice(1)) {
    if (value === prev + 1) {
      prev = value;
      continue;
    }
    parts.push(formatRange(start, prev, prefix));
    start = prev = value;
  }
  parts.push(formatRange(start, prev, prefix));
  return parts.join(",");
}

function formatRange(start, end, prefix) {
  const left = `${prefix}${String(start).padStart(3, "0")}`;
  const right = `${prefix}${String(end).padStart(3, "0")}`;
  return start === end ? left : `${left}-${right}`;
}

File: grok_worker/prompt_parser.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable


# 한글 조사(@S999가, @S001은 등)가 바로 붙는 경우도 참조 토큰으로 인식합니다.
# ASCII 영숫자/언더스코어가 뒤에 이어질 때만 토큰 연장으로 보고 제외합니다.
REFERENCE_TOKEN_RE = re.compile(r"@((?:[Ss]\d+)|(?:[1-5]))(?![A-Za-z0-9_])")


@dataclass
class PromptBlock:
    number: int
    tag: str
    body: str
    rendered_prompt: str
    raw: str
    references: list[str]


def _normalize_body(body: str) -> str:
    lines = [line.rstrip() for line in str(body or "").splitlines()]
    return "\n".join(lines).strip()


def _render_prompt(prefix: str, number: int, pad_width: int, body: str) -> tuple[str, str]:
    tag = f"{prefix}{str(number).zfill(max(3, int(pad_width or 3)))}"
    body_text = _normalize_body(body)
    return tag, f"{tag} Prompt : {body_text}"


def _normalize_prefix_aliases(prefix: str, extra_prefixes: Iterable[str] = ()) -> tuple[str, ...]:
    aliases: list[str] = []
    for token in (prefix, *(extra_prefixes or ())):
        normalized = str(token or "").strip().upper()
        if normalized and normalized not in aliases:
            aliases.append(normalized)
    if not aliases:
        aliases.append("S")
    return tuple(aliases)


def _normalize_prompt_chunk(chunk: str) -> str:
    lines = [line.rstrip() for line in str(chunk or "").splitlines()]
    return "\n".join(lines).strip()


def parse_prompt_blocks(
    raw_text: str,
    *,
    prefix: str = "S",
    pad_width: int = 3,
    separator: str = "|||",
    extra_prefixes: Iterable[str] = (),
) -> list[PromptBlock]:
    prefix_aliases = _normalize_prefix_aliases(prefix, extra_prefixes)
    prefix_pattern = "|".join(re.escape(token) for token in prefix_aliases)
    chunks = [part.strip() for part in str(raw_text or "").split(separator) if part.strip()]
    items: list[PromptBlock] = []
    for chunk in chunks:
        normalized_chunk = _normalize_prompt_chunk(chunk)
        if not normalized_chunk:
            continue

        lines = normalized_chunk.splitlines()
        first_line = str(lines[0] or "").strip() if lines else ""
        rest_lines = lines[1:]
        number = None
        body = ""
        tag = ""
        rendered = normalized_chunk

        labeled_match = re.match(
            rf"^\s*(((?:{prefix_pattern})\s*0*([1-9][0-9]*))(?:\s*>\s*(?:{prefix_pattern})\s*0*([1-9][0-9]*))?)\s*(?:PROMPT|프롬프트)\s*:\s*(.*)\s*$",
            first_line,
            re.IGNORECASE | re.DOTALL,
        )
        if labeled_match:
            number = int(labeled_match.group(3))
            prompt_spec = str(labeled_match.group(1) or "").strip()
            tag_head = prompt_spec.split(">", 1)[0].strip()
            prefix_match = re.match(r"^\s*([A-Za-z]+)", tag_head)
            display_prefix = str(prefix_match.group(1) or prefix).strip().upper() if prefix_match else str(prefix or "S").strip().upper()
            tag = f"{display_prefix}{str(number).zfill(max(3, int(pad_width or 3)))}"
            inline_body = str(labeled_match.group(5) or "").strip()
            body_parts = []
            if inline_body:
                body_parts.append(inline_body)
            if rest_lines:
                body_parts.append("\n".join(rest_lines).strip())
            body = _normalize_body("\n".join(part for part in body_parts if part.strip()))
            rendered = f"{prompt_spec.upper()} Prompt : {body}"
        else:
            inline_match = re.match(r"^\s*0*([1-9][0-9]*)\s*:\s*(.*)\s*$", first_line, re.DOTALL)
            if inline_match:
                number = int(inline_match.group(1))
                inline_body = str(inline_match.group(2) or "").strip()
                body_parts = []
                if inline_body:
                    body_parts.append(inline_body)
                if rest_lines:
                    body_parts.append("\n".join(rest_lines).strip())
                body = _normalize_body("\n".join(part for part in body_parts if part.strip()))
                tag, rendered = _render_prompt(prefix, number, pad_width, body)
            else:
                multi_match = re.match(r"^\s*0*([1-9][0-9]*)\s*:\s*$", first_line)
                if not multi_match:
                    continue
                number = int(multi_match.group(1))
                body = _normalize_body("\n".join(rest_lines))
                tag, rendered = _render_prompt(prefix, number, pad_width, body)

        if not number or not body:
            continue

        refs: list[str] = []
        seen_refs: set[str] = set()
        for match in REFERENCE_TOKEN_RE.finditer(rendered):
            token = str(match.group(1) or "").upper()
            if not token or token in seen_refs:
                continue
            seen_refs.add(token)
            refs.append(token)
        items.append(
            PromptBlock(
                number=number,
                tag=tag,
                body=body,
                rendered_prompt=rendered,
                raw=normalized_chunk,
                references=refs,
            )
        )
    items.sort(key=lambda item: item.number)
    return items


def load_prompt_blocks(
    path: Path,
    *,
    prefix: str = "S",
    pad_width: int = 3,
    separator: str = "|||",
    extra_prefixes: Iterable[str] = (),
) -> list[PromptBlock]:
    if not path.exists():
        return []
    raw = path.read_text(encoding="utf-8")
    return parse_prompt_blocks(raw, prefix=prefix, pad_width=pad_width, separator=separator, extra_prefixes=extra_prefixes)


def summarize_prompt_file(
    path: Path,
    *,
    prefix: str = "S",
    pad_width: int = 3,
    separator: str = "|||",
    extra_prefixes: Iterable[str] = (),
) -> str:
    items = load_prompt_blocks(path, prefix=prefix, pad_width=pad_width, separator=separator, extra_prefixes=extra_prefixes)
    if not items:
        return f"{path.name} | 프롬프트 없음"
    number_text = ",".join(f"{item.number:03d}" for item in items)
    return f"{path.name} | 총 {len(items)}개 | {number_text}"


def compress_numbers(numbers: Iterable[int], prefix: str = "") -> str:
    nums = sorted({int(n) for n in numbers if int(n) > 0})
    if not nums:
        return ""
    ranges: list[str] = []
    start = prev = nums[0]
    for value in nums[1:]:
        if value == prev + 1:
            prev = value
            continue
        ranges.append(_format_range(start, prev, prefix))
        start = prev = value
    ranges.append(_format_range(start, prev, prefix))
    return ",".join(ranges)


def _format_range(start: int, end: int, prefix: str = "") -> str:
    left = f"{prefix}{start:03d}"
    right = f"{prefix}{end:03d}"
    if start == end:
        return left
    return f"{left}-{right}"
